dbus-wrapper/swupdd-main.c

```diff
@@ -0,0 +1,985 @@
+
+	char const * const str_opts[] = {"url", "contenturl", "versionurl",
+					 "path", "format", "statedir", NULL};
+	char const * const bool_opts[] = {"list", "force", NULL};
+	char const * const int_opts[] = {"port", NULL};
+	r = bus_message_read_options(m, str_opts, bool_opts, int_opts, &args, ret_error);
+	if (r < 0) {
+		goto finish;
+	}
+
+	r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "s");
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Can't enter bundles container");
+		goto finish;
+	}
+	while ((r = sd_bus_message_read(m, "s", &bundle)) > 0) {
+		args = list_append_data(args, strdup(bundle));
+	}
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Can't read bundle name");
+		goto finish;
+	}
+	r = sd_bus_message_exit_container(m);
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Can't exit bundles container");
+		goto finish;
+	}
+
+	r  = run_swupd(METHOD_BUNDLE_ADD, args, context);
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Failed to run swupd command");
+		goto finish;
+	}
+
+	r = sd_bus_reply_method_return(m, "b", (r >= 0));
+
+finish:
+	list_free_list_and_data(args, free);
+	return r;
+}
+
+static int method_bundle_remove(sd_bus_message *m,
+				void *userdata,
+				sd_bus_error *ret_error)
+{
+	daemon_state_t *context = userdata;
+	int r = 0;
+	struct list *args = NULL;
+
+	r = check_prerequisites(context, ret_error);
+	if (r < 0) {
+		return r;
+	}
+
+	args = list_append_data(args, strdup(SWUPD_CLIENT));
+	args = list_append_data(args, strdup(_method_opt_map[METHOD_BUNDLE_REMOVE]));
+
+	char const * const str_opts[] = {"path", "url", "contenturl", "versionurl",
+					 "format", "statedir", NULL};
+	char const * const bool_opts[] = {"force", NULL};
+	char const * const int_opts[] = {"port", NULL};
+	r = bus_message_read_options(m, str_opts, bool_opts, int_opts, &args, ret_error);
+	if (r < 0) {
+		goto finish;
+	}
+
+	const char *bundle;
+	r = sd_bus_message_read(m, "s", &bundle);
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Can't read bundle name");
+		goto finish;
+	}
+	args = list_append_data(args, strdup(bundle));
+
+	r  = run_swupd(METHOD_BUNDLE_REMOVE, args, context);
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Failed to run swupd command");
+		goto finish;
+	}
+
+	r = sd_bus_reply_method_return(m, "b", (r >= 0));
+
+finish:
+	list_free_list_and_data(args, free);
+	return r;
+}
+
+static int method_cancel(sd_bus_message *m,
+			 void *userdata,
+			 sd_bus_error *ret_error)
+{
+	daemon_state_t *context = userdata;
+	int r = 0;
+	pid_t child;
+	int force;
+
+	child = context->child;
+	if (!child) {
+		sd_bus_error_set_errnof(ret_error, ECHILD, "No child process to cancel");
+		return -ECHILD;
+	}
+
+	r = sd_bus_message_read(m, "b", &force);
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Can't read 'force' option");
+		return r;
+	}
+
+	if (force) {
+		kill(child, SIGKILL);
+	} else {
+		kill(child, SIGTERM);
```

Comment:
This says swupd-client should probably start handling SIGTERM to be a little more graceful in shutdown.